### PR TITLE
[IMP]: emit libsass warnings through the logging system

### PR DESCRIPTION
**Pull request by xmo-odoo**
libsass is currently hard-coded to emit deprecation warnings to stderr (sass/libsass#3079), though explicit warnings can be intercepted (I could not say whether they *are* by the Python bindings though).

Since libsass 3.5 (released July 2017), extending compound selector is deprecated ([0][bc], [1][extendcompound]) triggering the following message:

    WARNING on line 6713, column 13 of stdin:
    Compound selectors may no longer be extended.
    Consider `@extend .form-control, :disabled` instead.
    See http://bit.ly/ExtendCompound for details.

This is both basically impossible to see when running "full" logs, and extremely spammy when only showing "warning"-level logs and running tours, as each tour will cause at least one assets compilation, triggering the warning if using a non-ancient libsass. Because it's emitted directly to stderr, it's also difficult to filter out.

This PR messes arounds with fd redirections (adding a helper for that, similar to `contextlib.redirect_stderr` but working at the fd level as that's what we need here) in order to capture the stderr output of libsass and re-emit it as a logging `warning` if non-empty, to make it both visible and filter-able.

* Intercepts to a temporary file, as we don't really know how much *stuff* libsass may output, currently it's way below PIPE_BUF (to say nothing of the actual pipe buffers) but we've no way to know that will always be the case, so using a pipe would require a separate thread to ensure the pipe output gets read and we don't risk clogging up the pipe (and blocking libsass entirely).
* Doesn't bother checking for content with `tell` beforehand, that seems unlikely to be a limiting factor: currently we always lseek + read, by first checking for content with `tell` we'd have a "best case" (libsass emitted no warnings) of lseek but a "worst case" (libsass emitted warnings) of lseek + lseek + read.
* Emitting a logging event rather than a `Deprecation` warning, an actual deprecation warning might be useful if we were processing the actual source files and could provide a precise source location to `warn_explicit`, but here we're dealing with a concatenated pile of junk (the warning is signaled on line 6713, even the largest bootstrap scss file is barely above 1Kloc) so shoving it through logging seems sufficient.

[bc]: https://sass-lang.com/documentation/breaking-changes/extend-compound
[extendcompound]: http://bit.ly/ExtendCompound

### CODE REVIEW

**Review by d-fence — APPROVED** [review]
LGTM
As you explained me offline, the libssas warning is split across multiple lines ad thus difficult to parse. That justify the fact of having one python warning.

**Comment by xmo-odoo** [comment]
To add to the commit message, after discussion with xdo: this should not have thread-safety issues, it's not a concern at all in worker mode, and in threaded mode... `libsass` does not seem to release the GIL, neither `PyEval_SaveThread` nor `Py_BEGIN_ALLOW_THREADS` seems used anywhere in the codebase. So essentially the only concurrency risk is thread-switching between the end of the `libsass.compile` code and the end of the context manager.

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security r+

**Comment by odony** [comment]
aren't we fixing the line that triggers the warning in `web_editor.frontend.scss`, instead of hiding it?

**Comment by xmo-odoo** [comment]
> Aren't we fixing the line that triggers the warning in `web_editor.frontend.scss`, instead of hiding it? or do we expect more similar warnings to be frequently triggered?

Well this patch is really properly surfacing the warning (as a warning). The side-effect is that it *can* be hidden using log-handler and friends, but the baseline of this patch is to actually show it.

I have no idea what the future plans of the Sass Committee are, and what might become deprecated in the future, and thus can't really talk about the long-term risks of new warnings in the future. But I think surfacing those warnings, if they happen, would be still be useful.

> I'm asking because the patch is rather invasive, and seems to count on a bit of luck for thread safety.

Yes, although I think the thread unsafety issues are relatively limited: worst case scenario would be for the current thread to be descheduled in the small windows around libsass' native calls and an other thread scheduled at that point to log stuff, leading to log records being shoved into the file being collected and thus attributed to libsass warnings.

> I'd rather just fix the SCSS, which might even be done in stable and in a backwards-compatible manner, IIUC?

Maybe. I have no idea what @qsm-odoo originally intended though, I'm rather confused by the explanations of the warning and it looks like the use of `@extend` basically never worked as intended. I'd expect

	  .o_wysiwyg_loader {
	      @extend .form-control:disabled;
	  }

to be intended as "`.o_wysiwyg_loader` should be styled like a disabled form control", but the warning seems to say that it never worked that way, and instead would be styled like a form control and a disabled element? The latter  So maybe just removing `:disabled` would be good enough. Alternatively, [it might never have worked at all because pseudo-classes were involved](https://stackoverflow.com/questions/17181837/sass-extend-with-pseudo-selectors).

**Comment by xmo-odoo** [comment]
@robodoo r-

**Comment by qsm-odoo** [comment]
Sorry, I was on leaves last week.

TL;DR replacing the extend by `@extend :disabled;` is perfect for master and probably ok for stable.

---

> Maybe. I have no idea what @qsm-odoo originally intended though, I'm rather confused by the explanations of the warning and it looks like the use of extend basically never worked as intended. I'd expect the code to be intended as ".o_wysiwyg_loader should be styled like a disabled form control"

To actually say "get the style of a form-control which is disabled", it would be `@extend .form-control, .form-control:disabled` but here it works as intended in our code base since we actually use the `o_wysiwyg_loader` class alongside the `form-control` class explicitly (except in one of our test which do not care about the style). So I did get the result I wanted but missed the fact the `.form-control` part was not really needed in my selector.

> So maybe just removing :disabled would be good enough

No, that would make it look like a normal form-control which is not what we want (and this would actually have no effect in our codebase since we already use the class).

**2 solutions:**

1)

```scss
@extend :disabled;
```
-> actually supposes we explicitly use the `form-control` class alongside the `o_wysiwyg_loader` one. If done in master, probably the best solution. If done in stable, just "breaks" the style of potential custo which do not use the `form-control` class explicitly (but we are talking about the loading effect of editor inputs on page load and if the custo did not use that class, the input was already ugly since my code was not written  `@extend .form-control, .form-control:disabled` as explained above...).

2)

```scss
    background-color: $input-disabled-bg;
```
-> copy paste of what the current extend does. Except for this

```scss
    // iOS fix for unreadable disabled content; see https://github.com/twbs/bootstrap/issues/11655.
    opacity: 1;
```

part which we do not really care about for this loading effect (and probably not needed at all since this iOS fix probably only concerns inputs which actually are disabled). Don't rely on the fact we use the form-control class explicitly, so better for stable version in that sense but do not get all the potential custom style of `.form-control:disabled` made by potential custo.

-> I would choose solution (1).

**Comment by xmo-odoo** [comment]
OK I'll create a PR to implement (1) in 14.0 then, that's probably fairly good.
